apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/SpannerConnection.cs

```diff
@@ -0,0 +1,313 @@
+            }
+
+        }
+        private void TrySetNewConnectionInfo(SpannerConnectionStringBuilder newBuilder) {
+            AssertClosed("change connection information.");
+            _connectionStringBuilder = newBuilder;
+        }
+
+        /// <summary>
+        /// </summary>
+        /// <returns></returns>
+        public SpannerBatchOperation BeginBatchOperation()
+        {
+            //This may be called while the connection is closed.
+            SpannerBatchOperation newBatchOperation = new SpannerBatchOperation(this);
+            if (Interlocked.CompareExchange(ref _activeBatchOperation, newBatchOperation, null)
+                != null) {
+                throw new InvalidOperationException("You may only create a single Batch operation at a time.");
+            }
+            return _activeBatchOperation;
+        }
+
+#if NET451
+        /// <inheritdoc />
+        public override void EnlistTransaction(System.Transactions.Transaction transaction)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc />
+        protected override DbProviderFactory DbProviderFactory
+        {
+            get { return SpannerProviderFactory.Instance; }
+        }
+#endif
+
+        /// <summary>
+        /// Provides options to customize how connections to Spanner are created
+        /// and maintained.
+        /// </summary>
+        public static ConnectionPoolOptions ConnectionPoolOptions
+        {
+            get { return ConnectionPoolOptions.Instance; }
+        }
+
+        /// <summary>
+        /// Returns the credential used for the connection, if set.
+        /// </summary>
+        public ITokenAccess Credential
+        {
+            get { return _connectionStringBuilder.Credential; }
+        }
+
+        /// <inheritdoc />
+        public override string ConnectionString
+        {
+            get { return _connectionStringBuilder.ToString(); }
+            set { TrySetNewConnectionInfo(new SpannerConnectionStringBuilder(value, _connectionStringBuilder.Credential)); }
+        }
+
+        /// <inheritdoc />
+        public override string Database
+        {
+            get { return _connectionStringBuilder.SpannerDatabase; }
+        }
+
+        /// <inheritdoc />
+        public override ConnectionState State
+        {
+            get { return _state; }
+        }
+
+        /// <inheritdoc />
+        public override void Open() {
+            if (IsOpen)
+            {
+                return;
+            }
+            OpenAsync(CancellationToken.None).Wait();
+        }
+
+        /// <inheritdoc />
+        public override async Task OpenAsync(CancellationToken cancellationToken) {
+            if (IsOpen) {
+                return;
+            }
+            _client = await ClientPool.AcquireClientAsync(_connectionStringBuilder.Credential);
+            _session =
```

Comment:
the parallel wait won't actually accomplish anything because the session call needs the client as well, but I agree that's not clear and it wouldnt hurt to wait in parallel as a good practice.
but multiple thread access should be guarded here.